### DISCUSSION TITLE
About Conditional Rendering

### DISCUSSION THREAD

**Discussion by noo-have** [solidjs/solid#1246] [Q&A] [answered]
I have a problem: when I write down such jsx：
```tsx
function Counter() {
  const [count, setCount] = createSignal(0)
  const inc = () => setCount(count() + 1)
  const dec = () => setCount(count() - 1)

  return (
    <div>
      <div>{count()}</div>
      <button type='button' onClick={() => inc()}>
        +
      </button>
      <button type='button' onClick={() => dec()}>
        -
      </button>
      {count() > 2 && <Child w={count()} />}
    </div>
  )
}

function Child(props: { w: number }) {
  console.log(1)
  return <div>{props.w}</div>
}
```
the `child` component will be re-rendered as `count` changes. (`console.log(1)` will be called again)

So I used `createMemo` for caching, but I found that the `child` component was called before `count` met the criteria.

I don't want to, so I write down a `cache function` like this:

```tsx
const useCache = <T,>(elm: () => T) => {
    let cache: T
    return () => (cache ??= elm())
}
let cache = useCache(() => <Child w={count()} />)
// {count() > 2 && <Child w={count()} />}  ====》 {count() > 2 && cache()}

```

Now the `child` component won't render in advance, and when count first meets the criteria, everything looks fine.

After rendering the `child` component for the first time, the `child` component disappears when I reduce count so that it is less than or equal to 2 (which is predictable), but once I make count greater than 2 again, the `child` component does not change as `count` increases: it looks like it has lost its responsiveness.

What shall I do?

**Comment by ryansolid** [comment, marked as answer]
This has to do with the fact that children get released when parents re-run so holding a reference doesn't help you. There are solves for that, but let's backpedal a moment.

The original code only will re-run when the component is re-mounted again. So 2 or less.. not mounted, 3 console.logs, 4 no log, 5 no log, back down to 2 unmounts but no logs, back up to 3, logs again.

So to be clear you want it to lazily be created but then stick around until Counter is unmounted? You could hide it with styling like `display: none` and then it never goes away and you don't have this problem. But if you want to go the unmount route you can do this I think:
```js
function useCache(fn) {
  let cache;
  let owner = getOwner();
  return () => cache ??= runWithOwner(owner, fn);
}
```
I generally caution here because this isn't as general purpose as it looks. Make sure you only use it in one location. DOM nodes can only be attached to the DOM in a single location.

**Reply by noo-have** [reply]
Thank you for your reply! My problem has been solved successfully! But what are getOwner and runWithOwner ? I don't see them in the API documentation.

**Reply by ryansolid** [reply]
Yeah I probably need to add a super advanced section. There are a few helpers that get you access to the reactive system execution internals. They are hard to explain without full understanding of Solid so we sort of need docs to catch up to them.
